ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/InitiatorBase.java

```diff
@@ -50,59 +49,53 @@
 
 import java.io.IOException;
 import java.security.PrivilegedExceptionAction;
-import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 public class InitiatorBase extends MetaStoreCompactorThread {
 
   static final private String COMPACTOR_THRESHOLD_PREFIX = "compactorthreshold.";
 
-  private List<CompactionResponse> initiateCompactionForMultiplePartitions(Table table,
-      Map<String, Partition> partitions, CompactionRequest request) {
-    List<CompactionResponse> compactionResponses = new ArrayList<>();
-    partitions.entrySet().parallelStream().forEach(entry -> {
-      try {
-        StorageDescriptor sd = CompactorUtil.resolveStorageDescriptor(table, entry.getValue());
-        String runAs = TxnUtils.findUserToRunAs(sd.getLocation(), table, conf);
-        CompactionInfo ci =
-            new CompactionInfo(table.getDbName(), table.getTableName(), entry.getKey(), request.getType());
-        ci.initiatorId = request.getInitiatorId();
-        ci.orderByClause = request.getOrderByClause();
-        ci.initiatorVersion = request.getInitiatorVersion();
-        if (request.getNumberOfBuckets() > 0) {
-          ci.numberOfBuckets = request.getNumberOfBuckets();
-        }
-        ci.poolName = request.getPoolName();
-        LOG.info(
-            "Checking to see if we should compact partition " + entry.getKey() + " of table " + table.getDbName() + "."
-                + table.getTableName());
-        CollectionUtils.addIgnoreNull(compactionResponses,
-            scheduleCompactionIfRequired(ci, table, entry.getValue(), runAs, false));
-      } catch (IOException | InterruptedException | MetaException e) {
-        LOG.error(
-            "Error occurred while Checking if we should compact partition " + entry.getKey() + " of table " + table.getDbName() + "."
-                + table.getTableName() + " Exception: " + e.getMessage());
-        throw new RuntimeException(e);
-      }
-    });
-    return compactionResponses;
-  }
-
-  public List<CompactionResponse> initiateCompactionForTable(CompactionRequest request, Table table, Map<String, Partition> partitions) throws Exception {
+  public void initialize() throws Exception {
+    super.init(new AtomicBoolean());
```

Comment:
Yeah, i have overridden init itself.

Reply:
The latest patch changes the behavior of `Initiator` forcing the latter to retrieve and set a `validTxnList` during initialization something that wasn't done before. I reviewed related code in `Initiator` class and the change appears safe but in general it is best to avoid such refactorings that implicitly change the behavior of the code.

@tarak271 Did you double check that there are no undesired effects from this change in behavior? If yes then we can proceed and merge this PR.

Reply:
@zabetak I have moved all the reused functions into CompactorUtil.java by making them static. So now initiator as well as manual compaction will re-use the functions and no inheritance between these two, so no issues w.r.t Initiator. Removed InitiatorBase as well. Please take a look. Thanks a lot for your time on this. Sorry that I made you review more number of times

Reply:
Moving the methods in `CompactorUtil` is good idea many thanks for doing that. I added some comments about the last refactoring in the respective places.